espresso/src/com.oracle.truffle.espresso/src/tools/aqua/concolic/ConcolicAnalysis.java

```diff
@@ -804,6 +805,51 @@ public void takeBranchPrimitive2(VirtualFrame frame, BytecodeNode bcn, int bci,
         trace.addElement(pc);
     }
 
+    @Override
+    public void takeBranchRef2(VirtualFrame frame, BytecodeNode bcn, int bci, int opcode, boolean takeBranch, StaticObject c1, StaticObject c2, Expression a1, Expression a2) {
+        if ((a1 == null) && (a2 == null)) {
+            return;
+        }
+
+        Expression expr = null;
+        Meta meta = bcn.getMeta();
+
+        // special case: cached Integer
+        if (meta.java_lang_Integer.equals(c1.getKlass()) &&
+                meta.java_lang_Integer.equals(c2.getKlass())) {
+
+            Expression e1 = c1 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c1, meta.java_lang_Integer_value), config.getConcolicIdx());
+            Expression e2 = c2 == null ? null : Annotations.annotation(
+                    AnnotatedVM.getFieldAnnotation(c2, meta.java_lang_Integer_value), config.getConcolicIdx());
+
+            if (e1 != null || e2 != null) {
+
+                int int1 = meta.java_lang_Integer_value.getInt(c1);
+                int int2 = meta.java_lang_Integer_value.getInt(c2);
+
+                e1 = e1 == null ? Expression.fromConstant(Types.INT, int1) : e1;
+                e2 = e2 == null ? Expression.fromConstant(Types.INT, int2) : e2;
+
+                expr = new ComplexExpression(BAND,
+                        new ComplexExpression(BVEQ, e1, e2),
+                        new ComplexExpression(BVLE, Expression.fromConstant(Types.INT, -128), e1),
+                        new ComplexExpression(BVLE, e1, Expression.fromConstant(Types.INT, 127)),
+                        // remaining not strictly necessary?
+                        new ComplexExpression(BVLE, Expression.fromConstant(Types.INT, -128), e2),
+                        new ComplexExpression(BVLE, e2, Expression.fromConstant(Types.INT, 127)));
+
+                if (!(int1 == int2 && -128 <= int1 && int1 <= 127)) {
```

Comment:
This check is only implemented on int1 while the expression affects int2. Is this mismatch intended?

Reply:
if int1 == int2 then what is tested afterwards holds for int1 and int2. I was more verbose in the symbolic constraint